loader/src/test/scala/com/snowplowanalytics/snowplow/storage/bigquery/loader/LoaderCacheSpec.scala

```diff
@@ -0,0 +1,101 @@
+ */
+package com.snowplowanalytics.snowplow.storage.bigquery
+package loader
+
+import java.time.Instant
+import cats.data.NonEmptyList
+import com.snowplowanalytics.iglu.client.ClientError
+import com.snowplowanalytics.iglu.client.resolver.LookupHistory
+import com.snowplowanalytics.iglu.core.{SchemaKey, SchemaVer}
+import com.snowplowanalytics.iglu.schemaddl.bigquery.{Field, Mode, Type}
+import com.snowplowanalytics.snowplow.badrows.FailureDetails
+
+import org.specs2.mutable.Specification
+
+class LoaderCacheSpec extends Specification {
+  import com.snowplowanalytics.snowplow.storage.bigquery.loader.LoaderCache._
+
+  sequential
+
+  "get should" >> {
+    "return None if the key is not in the cache" >> {
+      val __ = underlying.clear()
+      val key = SchemaKey("com.acme", "missing_key", "jsonschema", SchemaVer.Full(1, 0, 0))
+
+      get(key) shouldEqual None
+    }
+
+    "return the value if the key is in the cache" >> {
+      val __ = underlying.clear()
+      val key = SchemaKey("com.acme", "available_key", "jsonschema", SchemaVer.Full(1, 0, 0))
+      val value = Right(Field("com_acme_available_key_1", Type.String, Mode.Nullable))
+      val _ = put(key, value)
+      val result = get(key)
+
+      result.get._1 shouldEqual(value)
+    }
+  }
+
+  "put should" >> {
+    "return the value after inserting it in the cache" >> {
+      val __ = underlying.clear()
+      val key = SchemaKey("com.acme", "new_key", "jsonschema", SchemaVer.Full(1, 0, 0))
+      val value = Right(Field("com_acme_new_key_1", Type.String, Mode.Nullable))
+      val putResult = put(key, value)
+      val getResult = get(key)
+
+      putResult shouldEqual(value)
+      getResult.get._1 shouldEqual(value)
+    }
+
+    "override an existing key in the cache" >> {
+      val __ = underlying.clear()
+      val key = SchemaKey("com.acme", "existing_key", "jsonschema", SchemaVer.Full(1, 0, 0))
+      val oldValue = Right(Field("com_acme_existing_key_1", Type.String, Mode.Nullable))
+      val newValue = Right(Field("com_acme_existing_key_1", Type.String, Mode.Repeated))
+      val _ = put(key, oldValue)
+      val cacheSize = underlying.size
+
+      get(key).get._1 shouldNotEqual { put(key, newValue); get(key).get._1 }
+      { put(key, oldValue); put(key, newValue); underlying.size } shouldEqual cacheSize
+    }
+  }
+
+  "getOrLookup should" >> {
+    val key = SchemaKey("com.acme", "existing_key", "jsonschema", SchemaVer.Full(1, 0, 0))
+    val value = Right(Field("com_acme_existing_key_1", Type.String, Mode.Nullable))
+    val error = Left(NonEmptyList.one(FailureDetails.LoaderIgluError.IgluError(key, ClientError.ResolutionError(Map("1" -> LookupHistory(Set(), 1, Instant.now))))))
+
+    "get the value from cache if it's there" >> {
+      val __ = underlying.clear()
+      val _ = put(key, value)
+
+      getOrLookup(key)(error) shouldEqual(value)
+    }
+
+    "not execute f if the value is in the cache" >> {
```

Comment:
This does not test that `f` is not executed, only that it's not returned

Reply:
Good point.